bindata/network/ovn-kubernetes-ipsec/ovn-ipsec.yaml

```diff
@@ -0,0 +1,207 @@
+      priorityClassName: "system-node-critical"
+      # volumes in all containers:
+      # (container) -> (host)
+      # /etc/openvswitch -> /var/lib/openvswitch/etc - ovsdb system id
+      # /var/lib/openvswitch -> /var/lib/openvswitch/data - ovsdb data
+      # /run/openvswitch -> tmpfs - ovsdb sockets
+      # /env -> configmap env-overrides - debug overrides
+      containers:
+      # ovs-monitor-ipsec and libreswan daemons
+      - name: ovn-ipsec
+        image: "{{.OvnImage}}"
+        command:
+        - /bin/bash
+        - -c
+        - |
+          #!/bin/bash
+
+          function cleanup()
+          {
+            # Stop ovn-monitor-ipsec
+            /usr/share/openvswitch/scripts/ovs-ctl stop-ovs-ipsec
+            # Stop libreswan
+            /usr/libexec/ipsec/whack --shutdown
+            /sbin/ip xfrm policy flush
+            /sbin/ip xfrm state flush
+            /usr/sbin/ipsec --stopnflog
+            exit 0
+          }
+          trap cleanup SIGTERM
+
+          # Every time we restart the container, we will create a new key pair
+          echo "Configuring IPsec keys"
+
+          # We use the system-id as the CN. This is a requirement for OVN
+          cn=$(ovs-vsctl --retry -t 60 get Open_vSwitch . external-ids:system-id | tr -d "\"")
+
+          mkdir -p /etc/openvswitch/keys
+          pushd /etc/openvswitch/keys
+          umask 077 && openssl genrsa -out ipsec-privkey.pem 2048
+          openssl req -new -text \
+                  -extensions v3_req \
+                  -addext "subjectAltName = DNS:${cn}" \
+                  -subj "/C=US/O=ovnkubernetes/OU=kind/CN=${cn}" \
+                  -key ipsec-privkey.pem \
+                  -out ipsec-req.pem
+
+          csr_64=$(cat ipsec-req.pem | base64 | tr -d "\n")
+
+          # The signer controller does not allow re-signing a key. We will
+          # delete the old key to be sure it is not there
+          kubectl delete --ignore-not-found=true csr/$(hostname)
+
+          cat <<EOF | kubectl apply -f -
+          apiVersion: certificates.k8s.io/v1
+          kind: CertificateSigningRequest
+          metadata:
+            name: $(hostname)
+          spec:
+            request: ${csr_64}
+            signerName: network.openshift.io/signer
+            usages:
+            - ipsec tunnel
+          EOF
+
+          until [ ! -z $(kubectl get csr/$(hostname) -o jsonpath='{.status.certificate}' 2>/dev/null) ]
+          do
+            ((counter++))
+            if [ $counter -gt 60 ];
+            then
+                    echo "Unable to sign certificate after $counter seconds"
+                    exit 1
+            fi
+          done
+
+          kubectl get csr/$(hostname) -o jsonpath='{.status.certificate}' | base64 -d | openssl x509 -outform pem -text -out ipsec-cert.pem
+
+          kubectl delete csr/$(hostname)
+
+          cat /signer-ca/tls.crt | openssl x509 -outform pem -text > /etc/openvswitch/keys/ipsec-cacert.pem
```

Comment:
All of these writes need to be atomic (moving a symlink), because we don't know when ovs-vswitchd will read the certificate.

Reply:
In other words, write this to `/etc/openvswitch/keys/ipsec-cert-NNNN.pem` and then set a symlink

Reply:
Yeah, I will do this.

Reply:
Just checking this, the way "ovs-monitor-ipsec" works is that it monitors OVSDB for changes and acts upon them. In the case of certificates, when it sees the certificate has been added, it will import that cert into NSS. It will then only update the cert if the location changes or if "ovs-monitor-ipsec" is restarted. As we only update the keys when restarting the pod, neither of these conditions will trigger so we are safe to do this here.

Reply:
@abhat @squeed Updated this